pysph/tools/ipy_viewer.py

```diff
@@ -356,150 +357,206 @@ def _frame_handler(self, change):
+                sct = self._scatters[array_name] = self._scatter_ax.scatter(
                     temp_data[array_name].x,
                     temp_data[array_name].y,
-                    c=getattr(
-                            temp_data[array_name],
-                            pa_widgets.scalar.value
-                    ),
                     s=pa_widgets.scalar_size.value,
-                    cmap=pa_widgets.scalar_cmap.value,
                 )
-        self._legend_handler(None, manual=True)
+                c = getattr(
+                        temp_data[array_name],
+                        pa_widgets.scalar.value
+                )
+                c = c + abs(np.min(c))
+                cmap = pa_widgets.scalar_cmap.value
+                colormap = getattr(mpl.cm, cmap)
+                sct = self._scatters[array_name]
+                sct.set_facecolors(colormap(c*1.0/np.max(c)))
+        self._scatter_ax.axis('equal')
+        self._legend_handler(None)
 
-        # _vectors #
+    def _plot_vectors(self):
 
-        for vct in self._vectors.values():
-            if vct in self._scatter_ax.collections:
-                self._scatter_ax.collections.remove(vct)
+        try:
+            temp_data = self.get_frame(self._widgets.frame.value)
+            temp_data = temp_data['arrays']
+            self.figure.delaxes(self._vector_ax)
+            self._vector_ax = self.figure.add_axes(
+                self._scatter_ax.get_position(),
+                frameon=False
+                )
+            self._vector_ax.get_xaxis().set_visible(False)
+            self._vector_ax.get_yaxis().set_visible(False)
+            self._vectors = {}
+
+            for array_name in self._widgets.particles.keys():
+                if self._widgets.particles[array_name].vector.value != '':
+                    pa_widgets = self._widgets.particles[array_name]
+                    temp_data = temp_data[array_name]
+                    x = temp_data.x
+                    y = temp_data.y
+                    v1 = getattr(
+                        temp_data,
+                        pa_widgets.vector.value.split(",")[0]
+                        )
+                    v2 = getattr(
+                        temp_data,
+                        pa_widgets.vector.value.split(",")[1]
+                        )
+                    vmag = (v1**2 + v2**2)**0.5
+                    self._vectors[array_name] = self._vector_ax.quiver(
+                        x,
+                        y,
+                        v1,
+                        v2,
+                        vmag,
+                        scale=pa_widgets.vector_scale.value,
+                        width=(pa_widgets.vector_width.value)/10000,
+                    )
+            self._vector_ax.set_xlim(self._scatter_ax.get_xlim())
+            self._vector_ax.set_ylim(self._scatter_ax.get_ylim())
+
+        except AttributeError:
```

Comment:
This try except seems rather strange to me and I don't like it.  It is not at all obvious why there should be an attribute error and why it is OK to just pass on it? Can you please elaborate?

Reply:
So, the vector widget takes input "var1,var2" where var1 and var2 should each be an output property of the respective particlearray (say "u,v" or "au,av"), in any other case (say the input is "foo,foo1"), an AttributeError is encountered and to circumvent that, I added the try-except block. `pass` because on entering any invalid input, it doesn't do anything. If need be, I can add a TextArea widget at the bottom to display all such errors in a user-friendly language.

Reply:
My issue with this is that the try block should enclose the smallest possible code where you anticipate the error and not a blanket one where someone is wondering why you need to catch an exception for all of that code.  Please modify it along these lines.